### DISCUSSION TITLE
Question about fields coming from _source

### DISCUSSION THREAD

**Discussion by niek1992** [jertel/elastalert2#787] [Q&A] [answered]
Hi All,

Currently doing my thesis on handling alerts properly coming from SIEM. using Wazuh and ElastAlert2 to periodically query ES for fields. One of the requirements is filtering/muting an alert. 

Alert Manager is perfect for that, however i'd like to have as much fields available in Alert Manager for filtering/muting. Currently i'm working with a (not yet finished and tested) Jinja2 template that somewhat does the job where i have several types of alerts that could come in (talking about Windows alerts) and display the various fields these alerts have. However if one of the fields isn't populated or undefined, the Jinja2 template throws an exception and the rule is disabled. 

Now one way to make this completely bulletproof is to write a huge Jinja2 template with a lot of checks if a variable is not undefined e.g. `{%- if _source is not undefined -%}` and making use of `|default('<NaN>')` but this process becomes quite tedious and (in my case) error prone. 

Am wondering if it's possible to simply push the entire alert through with all the fields + values. I don't care about how inefficient this is and if Alert Manager is able to keep up with processing these alerts as its for a Proof of Concept.

I've read the docs numerous times to get a better grasp on how i should configure this but as English isn't my native language (and have a hard time understanding without detailed examples) i haven't been able to get it working.

Thanks in advance for replying, i've been trying to tackle this issue for quite a while now. Getting frustrated.

Niek

**Comment by jertel** [comment]
You mentioned `Alert Manager` several times but this project is specifically for the `ElastAlert 2` server side alerting engine. There is no alert management included within this project. I'm not sure how to assist.

**Comment by niek1992** [comment, marked as answer]
I know this project is specifically for `ElastAlert 2` even though i've mentioned `Alertmanager` several times. Maybe i've worded things wrong in my original post, but what i'm trying to do is:

- Have ElastAlert 2 alert on many (if not all) alerts in my ElasticSearch cluster (works);
- Configure `Alertmanager` as a `alerter` and use `Alertmanager` for further processing (sort of works*);
- Have all the available fields of the alert `ElastAlert 2` finds in ElasticSearch available in Alertmanager (not yet accomplished).

What i ment with sort of works is, that the connection between `ElastAlert 2` and `Alertmanager` works fine. Alerts are being forwarded near-realtime. I am however lacking various fields i'd like to filter on in `Alertmanager`. I know this project only entails `ElastAlert 2`, and since i'm trying to configure the `Alertmanager` alerter which is part of this project i think i'm in the right location to ask this question. Might've worded it unclear. 

What i'm hoping was that i am able to do something like:

```alert_text_type: alert_text_jinja
alert_text: |
  Complete Wazuh alert: {{_source}}
  ```
  
  or
  
  ```
  alertmanager_fields:
    source: _source
  ```
  And have every (nested) field mapped to a key/value pair.

If you want i can give you a redacted example of an ElasticSearch log entry i'd like to have all fields being added to the alert in `Alertmanager`?

**Reply by jertel** [reply]
You can write an ElastAlert 2 [enhancement](https://elastalert2.readthedocs.io/en/latest/recipes/adding_enhancements.html) which will either:
 1) split out all fields from the source document and send each as a key/value pair, or
 2) convert the document to JSON and send that as a single field.

**Reply by niek1992** [reply]
> split out all fields from the source document and send each as a key/value pair

Sounds like the only feasible option if i want to retain filtering options in `Alertmanager`. Option `#2` wouldn't work because a single field (while unique) can't be filtered on individual values.

I think i'll need to read up on writing `enhancement`'s, as i have no idea where to begin. The examples i've found aren't helpful either.. Thanks anyway, will mark this question as solved.

**Reply by niek1992** [reply]
Sorry to bother you again, as i've just come around to do some coding to get this to work. How would i add the key/value pairs to a match without setting `match` fields in an `Enhancement` and adding them to the alert manually? I know i can do:
```
def process(self, match):
    for value in match.items():
        match[str(key)] = str(value)`
```
But that requires me to manually add the keys as a label. Is there a way to add those to the alert dynamically? Looked through the [code](https://github.com/jertel/elastalert2/blob/478eb6e9be25b330e4db7e7f56e45ab97f637294/elastalert/alerters/alertmanager.py#L47-L52) but afaik that won't allow me to add labels to an alert. Even if i enable `run_enhancements_first` it won't help me in this case. 

Ideally i would like to add some program logic to the `elastalert` rule, specifically for `alertmanager_fields` (or `alertmanager_labels`). But as far as i could tell that's only possible in `alert_subject` and `alert_text` with jinja templating.

Thanks for your time.

**Reply by jertel** [reply]
In the enhancement, populate `rule['alertmanager_fields']` using the keys from the match items.

**Reply by niek1992** [reply]
Could you point me into the right direction? I'm stuck and currently have the following Enhancement (goes one nested dict deep as of now):

```py
class TestAlertEnhancement(BaseEnhancement):
    def process(self,match):
        alertManagerFields = {}
        for key,value in match.items():
            # Check if it is an dict
            if isinstance(value, dict):
                # looping through nested dict
                for key2,value2 in value.items():
                    alertManagerFields.update({key2:value2})
            alertManagerFields.update({key:value})
        self.rule['alertmanager_fields'] = alertManagerFields   
```

When debugging this locally (by manually supplying a `match` object it throws no errors. However when running in live environment nothing is being added to `alertmanager_fields`. Initially i thought it was caused by the `elastalert rule` as i had `alertmanager_fields` defined there. However after i commenting it out in that rule, still nothing was being added.

**Reply by jertel** [reply]
Use `print` or log statements to debug operational flow during the live testing. 

As far as the code you provided, it appears you are trying to add both the key and value into the field list. But the `alertmanager_fields` option is supposed to be a list of label:fieldname tuples. Not the actual value in the field. If you want to do it that way you might as well populate the `alertmanager_labels` option directly and not mess with the `alertmanager_fields` option.

**Reply by niek1992** [reply]
I have limited Python skills (and obviously code skills as well). Either way i want those key/value pairs to end up in alerts in Alert manager. 

> you might as well populate the alertmanager_labels option directly and not mess with the alertmanager_fields option.

I'd prefer a simple solution indeed. Is it a matter of replacing `self.rule['alertmanager_fields']` with `self.rule['alertmanager_labels']`? Because looking at the documentation for `alertmanager_labels` makes me question if it would work as it's expecting values that match `^[a-zA-Z_][a-zA-Z0-9_]*$` regex. For example `192.168.0.18` does not match said regex i think..

**Reply by jertel** [reply]
![image](https://user-images.githubusercontent.com/16639611/168109779-789585f5-3712-45ef-8b08-ebd3e4305a4d.png)

It says _keys_ must match the regex, not values. Regardless, you will likely need to replace non-compliant chars with underscores in those labels, in order to satisfy alertmanager requirements. 

I've never used this alerter, so I am going completely off of the code in that alerter, and the docs.
